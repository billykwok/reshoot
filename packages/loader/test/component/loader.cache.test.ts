import { describe, beforeAll, afterAll, test, expect } from '@jest/globals';
import path from 'path';
import compile from './compiler';
import createMemfs from './createMemfs';
import { mockBase64, mockMetroHash } from './mock';

describe('Basic component test for @reshoot/loader', () => {
  beforeAll(() => {
    jest.doMock('metrohash', mockMetroHash('tEsThAsH'));
    jest.doMock('../../src/util/base64', mockBase64('data/of=someimage'));
    jest.doMock('../../src/util/version', () => ({
      __esModule: true,
      default: '1.2.3',
    }));
  });

  afterAll((done) => {
    jest.resetModules();
    setTimeout(() => ((done as unknown) as () => void)(), 1000).unref();
  });

  test('should emit cached files and output successfully with default options', async () => {
    const memfs = createMemfs();
    jest.doMock('fs-extra', () => ({
      __esModule: true,
      readJson: async (filepath: string) => {
        const content = await memfs.promises.readFile(filepath);
        return JSON.parse(content.toString()) as unknown;
      },
      readFile: async (filepath: string) => {
        return await memfs.promises.readFile(filepath);
      },
      outputFile: async (filepath: string, data: Buffer | string) => {
        await memfs.promises.mkdir(path.dirname(filepath), { recursive: true });
        await memfs.promises.writeFile(filepath, data);
      },
      outputJson: async (filepath: string, json: any) => {
        await memfs.promises.mkdir(path.dirname(filepath), { recursive: true });
        await memfs.promises.writeFile(filepath, JSON.stringify(json));
      },
    }));
    const actual1 = await compile(
      memfs,
      '../../../../__fixtures__/test-image.jpg',
      { outputPath: '/images/', cache: true }
    );
    const images1 = await memfs.promises.readdir(
      path.resolve(__dirname, './images')
    );
    expect(images1).toEqual(
      expect.arrayContaining([
        'tEsThAsH-480.jpg',
        'tEsThAsH-640.jpg',
        'tEsThAsH-840.jpg',
        'tEsThAsH-1080.jpg',
        'tEsThAsH-480.webp',
        'tEsThAsH-640.webp',
        'tEsThAsH-840.webp',
        'tEsThAsH-1080.webp',
      ])
    );
    expect(images1).toHaveLength(8);
    expect(actual1).toMatchSnapshot();

    const cachePath = path.resolve(
      __dirname,
      '../../../../node_modules/.cache/@reshoot/loader/1.2.3/production/tEsThAsH'
    );
    const cachedOutput = await memfs.promises.readFile(
      path.join(cachePath, 'stats.json')
    );
    expect(cachedOutput.toString()).toMatchSnapshot();
    const cachedFiles = await memfs.promises.readdir(cachePath);
    expect(cachedFiles).toEqual(
      expect.arrayContaining([
        'stats.json',
        '480.jpg',
        '640.jpg',
        '840.jpg',
        '1080.jpg',
        '480.webp',
        '640.webp',
        '840.webp',
        '1080.webp',
      ])
    );

    const actual2 = await compile(
      memfs,
      '../../../../__fixtures__/test-image.jpg',
      { outputPath: '/images/', cache: true }
    );
    const images2 = await memfs.promises.readdir(
      path.resolve(__dirname, './images')
    );
    expect(images2).toEqual(
      expect.arrayContaining([
        'tEsThAsH-480.jpg',
        'tEsThAsH-640.jpg',
        'tEsThAsH-840.jpg',
        'tEsThAsH-1080.jpg',
        'tEsThAsH-480.webp',
        'tEsThAsH-640.webp',
        'tEsThAsH-840.webp',
        'tEsThAsH-1080.webp',
      ])
    );
    expect(images2).toHaveLength(8);
    expect(actual2).toMatchSnapshot();
    memfs.reset();
  }, 30000);
});
