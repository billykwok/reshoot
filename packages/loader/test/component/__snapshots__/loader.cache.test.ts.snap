// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic component test for @reshoot/loader should emit cached files and output successfully with default options 1`] = `"export default {\\"id\\":\\"tEsThAsH\\",\\"src\\":__webpack_public_path__+\\"tEsThAsH-1120.jpg\\",\\"width\\":4774,\\"height\\":3024,\\"srcSet\\":__webpack_public_path__+\\"tEsThAsH-640.jpg 640w,\\"+__webpack_public_path__+\\"tEsThAsH-1440.jpg 1440w\\",\\"aspectRatio\\":0.6334,\\"placeholder\\":\\"data:image/jpeg;base64,data/of=someimage\\",\\"color\\":\\"#283838\\",\\"sources\\":[{\\"type\\":\\"image/avif\\",\\"srcSet\\":__webpack_public_path__+\\"tEsThAsH-640.avif 640w,\\"+__webpack_public_path__+\\"tEsThAsH-1440.avif 1440w,\\"+__webpack_public_path__+\\"tEsThAsH-1120.avif 1120w\\"},{\\"type\\":\\"image/webp\\",\\"srcSet\\":__webpack_public_path__+\\"tEsThAsH-640.webp 640w,\\"+__webpack_public_path__+\\"tEsThAsH-1440.webp 1440w,\\"+__webpack_public_path__+\\"tEsThAsH-1120.webp 1120w\\"}]};"`;

exports[`Basic component test for @reshoot/loader should emit cached files and output successfully with default options 2`] = `"{\\"output\\":\\"export default {\\\\\\"id\\\\\\":\\\\\\"tEsThAsH\\\\\\",\\\\\\"src\\\\\\":__webpack_public_path__+\\\\\\"tEsThAsH-1120.jpg\\\\\\",\\\\\\"width\\\\\\":4774,\\\\\\"height\\\\\\":3024,\\\\\\"srcSet\\\\\\":__webpack_public_path__+\\\\\\"tEsThAsH-640.jpg 640w,\\\\\\"+__webpack_public_path__+\\\\\\"tEsThAsH-1440.jpg 1440w\\\\\\",\\\\\\"aspectRatio\\\\\\":0.6334,\\\\\\"placeholder\\\\\\":\\\\\\"data:image/jpeg;base64,data/of=someimage\\\\\\",\\\\\\"color\\\\\\":\\\\\\"#283838\\\\\\",\\\\\\"sources\\\\\\":[{\\\\\\"type\\\\\\":\\\\\\"image/avif\\\\\\",\\\\\\"srcSet\\\\\\":__webpack_public_path__+\\\\\\"tEsThAsH-640.avif 640w,\\\\\\"+__webpack_public_path__+\\\\\\"tEsThAsH-1440.avif 1440w,\\\\\\"+__webpack_public_path__+\\\\\\"tEsThAsH-1120.avif 1120w\\\\\\"},{\\\\\\"type\\\\\\":\\\\\\"image/webp\\\\\\",\\\\\\"srcSet\\\\\\":__webpack_public_path__+\\\\\\"tEsThAsH-640.webp 640w,\\\\\\"+__webpack_public_path__+\\\\\\"tEsThAsH-1440.webp 1440w,\\\\\\"+__webpack_public_path__+\\\\\\"tEsThAsH-1120.webp 1120w\\\\\\"}]};\\",\\"filenames\\":[\\"1120.jpg\\",\\"640.jpg\\",\\"1440.jpg\\",\\"640.avif\\",\\"1440.avif\\",\\"1120.avif\\",\\"640.webp\\",\\"1440.webp\\",\\"1120.webp\\"]}"`;

exports[`Basic component test for @reshoot/loader should emit cached files and output successfully with default options 3`] = `"export default {\\"id\\":\\"tEsThAsH\\",\\"src\\":__webpack_public_path__+\\"tEsThAsH-1120.jpg\\",\\"width\\":4774,\\"height\\":3024,\\"srcSet\\":__webpack_public_path__+\\"tEsThAsH-640.jpg 640w,\\"+__webpack_public_path__+\\"tEsThAsH-1440.jpg 1440w\\",\\"aspectRatio\\":0.6334,\\"placeholder\\":\\"data:image/jpeg;base64,data/of=someimage\\",\\"color\\":\\"#283838\\",\\"sources\\":[{\\"type\\":\\"image/avif\\",\\"srcSet\\":__webpack_public_path__+\\"tEsThAsH-640.avif 640w,\\"+__webpack_public_path__+\\"tEsThAsH-1440.avif 1440w,\\"+__webpack_public_path__+\\"tEsThAsH-1120.avif 1120w\\"},{\\"type\\":\\"image/webp\\",\\"srcSet\\":__webpack_public_path__+\\"tEsThAsH-640.webp 640w,\\"+__webpack_public_path__+\\"tEsThAsH-1440.webp 1440w,\\"+__webpack_public_path__+\\"tEsThAsH-1120.webp 1120w\\"}]};"`;
